### ISSUE TITLE
Encoding logic

### ISSUE THREAD

**Issue by magnumripper** [lakiw/pcfg_cracker#13]
I'm not quite sure how your current code works and anyway this is just food for thought and/or brainstorming:

If I were you I'd always use some version of Unicode internally (I often use Perl for encoding hacks, using whatever version of Unicode it has internally - I don't even have to know which it is), and use this logic for reading (per *line* which helps a lot when encountering mixed-encoding files - they are very common):

* Read line/string as "binary" - no encoding assumed.

* Add some kind of `valid_utf8(string)` function that can tell if UTF-8 is used or not.  (this check is trivial/quick and far more accurate than one would think intuitively). See for example JtR's `valid_utf8()` function in its' `unicode.c`. Here's the comments for that function (from `unicode.h`)
```c
/*
 * Check if a string is valid UTF-8.  Returns true if the string is valid
 * UTF-8 encoding, including pure 7-bit data or an empty string.
 *
 * The probability of a random string of bytes which is not pure ASCII being
 * valid UTF-8 is 3.9% for a two-byte sequence, and decreases exponentially
 * for longer sequences.  ISO/IEC 8859-1 is even less likely to be
 * mis-recognized as UTF-8:  The only non-ASCII characters in it would have
 * to be in sequences starting with either an accented letter or the
 * multiplication symbol and ending with a symbol.
 *
 * returns   0 if data is not valid UTF-8
 * returns   1 if data is pure ASCII (which is obviously valid)
 * returns > 1 if data is valid and in fact contains UTF-8 sequences
 *
 * Actually in the last case, the return is the number of proper UTF-8
 * sequences, so it can be used as a quality measure. A low number might be
 * a false positive, a high number most probably isn't.
 */
```
You could speed it up by just having it return true or false. Perhaps Python features some version of this check already for that matter, or you can just try decoding binary to UTF-8 with a try/catch.

Now, here's the logic for each line/string of binary data:

1. Optionally check if current string is double[or more]-encoded UTF-8. This check is simple and nearly 100% accurate, me thinks. It goes like this (pseudo code, or is it C 😆):
```c
	while (valid_utf8(string)) {
		new = decode_utf8(string);
		if (!valid_utf8(new))
			break;
		string = new;
	}
```
I've had LOTS of use of this and similar code is included in JtR's `cprepair` utility. Note that after the above, we continue with *next step*.

2. Check if current string conforms to UTF-8. If it is, we're done (possibly after converting to whatever other internal Unicode encoding we're using).

3. Failing the previous step, fall back to the defined legacy codepage (eg. CP1252) but convert it to whatever internal Unicode encoding we're using.

4. Optionally (and if so, included in step 3 above) you could even have a third encoding alternative like in my `cprepair` utility from JtR: A second legacy codepage, eg. CP737 or CP1251, that will kick in if we *have* a legacy codepage but there's *none* of `[a-zA-Z]` in the string. I've had excellent results with that in some cases.

Now, we have the internal database in known, proper Unicode after a best-effort conversion of possibly mixed input. For **output** encoding, I'd personally only support UTF-8 but you could just as well have it configurable. Although then you need a try/catch in case the word can't be expressed with the chosen legacy output encoding, and suppress such words.

**Comment by lakiw** [comment]
Hmm, that's good advice. I'll need to look into that more when I get a chance since character encoding still causes me problems.

**Comment by lakiw** [comment]
I implemented a variation of the above logic you proposed to v4.3 which has now been released. It attempts to re-encode strings to identify any unicode problems, as well as do things like handle $HEX[] encoded strings. I think there is value in allowing other code-pages vs. converting everything to UTF-8. Now eventually I'd like to add the ability to translate from one code-page to another, but that probably should go in whatever "mythical" ruleset updater tool that I've been talking about writing for the last 10 years but have never gotten around to coding ;p

Long story short, thank you once again for your comments and suggestions. As I mentioned, I have a variation of this logic now, but I'm sure that dealing with encoding issues will be a never-ending problem for this toolset. I'm going to close this issue for now, but if i get motivated I may open another one for additional encoding cleanup tasks.
